src/relations/mysql.py

```diff
@@ -0,0 +1,146 @@
+
+        Args:
+            username: The mysql username
+
+        Returns:
+            a string representing the password for the mysql user
+        """
+        peer_databag = self.charm._peers.data[self.charm.app]
+
+        if peer_databag.get(f"{username}_password"):
+            return peer_databag.get(f"{username}_password")
+
+        password = generate_random_password(PASSWORD_LENGTH)
+        peer_databag[f"{username}_password"] = password
+
+        return password
+
+    def _on_leader_elected(self, _) -> None:
+        """Handle the leader elected event.
+
+        Retrieves relation data from the peer relation databag and copies
+        the relation data into the new leader unit's databag.
+        """
+        # Skip if the charm is not past the setup phase (config-changed event not executed yet)
+        if not self.charm._is_peer_data_set:
+            return
+
+        relation_data = json.loads(
+            self.charm._peers.data[self.charm.app].get("mysql_relation_data", "{}")
+        )
+
+        for relation in self.charm.model.relations.get(LEGACY_MYSQL, []):
+            relation_databag = relation.data
+
+            # Copy relation data into the new leader unit's databag
+            for key, value in relation_data.items():
+                if relation_databag[self.charm.unit].get(key) != value:
+                    relation_databag[self.charm.unit][key] = value
+
+            # Assign the cluster primary's address as the database host
+            primary_address = self.charm._mysql.get_cluster_primary_address().split(":")[0]
+            relation_databag[self.charm.unit]["host"] = primary_address
+
+    def _on_mysql_relation_created(self, event: RelationCreatedEvent) -> None:
+        """Handle the legacy 'mysql' relation created event.
+
+        Will set up the database and the scoped application user. The connection
+        data (relation data) is then copied into the peer relation databag (to
+        be copied over to the new leader unit's databag in case of a new leader
+        being elected).
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        # Wait until on-config-changed event is executed
+        # (wait for root password to have been set)
+        if not self.charm._is_peer_data_set:
+            event.defer()
+            return
+
+        logger.warning("DEPRECATION WARNING - `mysql` is a legacy interface")
+
+        username = self.charm.config.get("mysql-interface-user")
+        database = self.charm.config.get("mysql-interface-database")
+
+        if not username or not database:
+            event.defer()
+            return
+
+        # Only execute if the application user does not exist
+        if self.charm._mysql.does_mysql_user_exist(username, "%"):
+            return
+
+        password = self._get_or_set_password_in_peer_databag(username)
+
+        self.charm._mysql.create_application_database_and_scoped_user(
```

Comment:
do you need to catch `MySQLCreateApplicationDatabaseAndScopedUserError` ? Or is it better to error out here? If its better to error out ignore me :)